### DISCUSSION TITLE
Query compilation

### DISCUSSION THREAD

**Discussion by noisywiz** [sqlalchemy/sqlalchemy#6104] [Usage Questions] [answered]
**Describe the bug**
My query:

```py
query = select([Department], limit=limit, offset=offset).where(
    Department.department_id.in_([x.department_id for x in permitted_departments]))
```

turns into:

```sql
SELECT department.department_id, department.department_name 
FROM department 
WHERE department.department_id IN ([POSTCOMPILE_department_id_1]) 
```

with error:

```
ERROR: syntax error at or near "["
```

**Expected behavior**

```sql
SELECT department.department_id, department.department_name 
FROM department 
WHERE department.department_id IN (%(department_id_1)s)
```

**Versions.**
 - OS: Debian buster
 - Python: 3.8
 - SQLAlchemy: 1.4.1
 - Database: Postgres

**Comment by CaselIT** [comment]
Hi,
How are you using the query? What you have posted seems like `str(query)`. That will not give you valid sql in the general case.

**Comment by noisywiz** [comment]
> Hi,
> How are you using the query? What you have posted seems like `str(query)`. That will not give you valid sql in the general case.

Hi, it's debug  print `query.compile(dialect=postgresql.dialect()))`. Actualy I fetch this with `databases.Database.fetch_all` method (https://pypi.org/project/databases/) and I'm fine with `SQLAlchemy==1.3.23`

**Comment by CaselIT** [comment, marked as answer]
Hi,

In 1.4 how the query is complied has changed to enable the caching of the generated strings.
To enable compilation if the query in a more complete way, you can use: `compile(dialect=dialect(), compile_kwargs={"literal_binds"=True, "render_postcompile"=True})` 

Note that non all datatypes support literal compile with `"literal_binds"=True`, so it may error on more complex types

**Comment by CaselIT** [comment]
@noisywiz do you have another more specific question or can we move this to a discussion?

**Comment by noisywiz** [comment]
> @noisywiz do you have another more specific question or can we move this to a discussion?

Yes close it. Thanks for help!
